mesonbuild/mintro.py

```diff
@@ -125,7 +131,153 @@ def list_target_files(target_name, coredata, builddata):
+                if comp is None:
+                    return
+                for optname in comp.base_options:
+                    if optname in self.coredata.base_options:
+                        continue
+                    oobj = compilers.base_options[optname]
+                    self.coredata.base_options[optname] = oobj
+
+    def func_project(self, node, args, kwargs):
+        if len(args) < 1:
+            raise InvalidArguments('Not enough arguments to project(). Needs at least the project name.')
+
+        proj_langs = self.flatten_args(args[1:])
+
+        if os.path.exists(self.option_file):
+            oi = optinterpreter.OptionInterpreter(self.subproject)
+            oi.process(self.option_file)
+            self.coredata.merge_user_options(oi.options)
+
+        def_opts = kwargs.get('default_options', [])
+        if isinstance(def_opts, mparser.ArrayNode):
+            def_opts = [x.value for x in def_opts.args.arguments]
+
+        self.project_default_options = mesonlib.stringlistify(def_opts)
+        self.project_default_options = cdata.create_options_dict(self.project_default_options)
+        self.default_options.update(self.project_default_options)
+
+        if not self.is_subproject() and 'subproject_dir' in kwargs:
+            spdirname = kwargs['subproject_dir']
+            if isinstance(spdirname, str):
+                self.subproject_dir = spdirname
+        if not self.is_subproject():
+            subprojects_dir = os.path.join(self.source_root, self.subproject_dir)
+            if os.path.isdir(subprojects_dir):
+                for i in os.listdir(subprojects_dir):
+                    if os.path.isdir(os.path.join(subprojects_dir, i)):
+                        self.do_subproject(i)
+
+        self.coredata.init_backend_options(self.backend.name)
+        options = {k: v for k, v in self.environment.cmd_line_options.items() if k.startswith('backend_')}
+
+        self.coredata.set_options(options)
+        self.add_languages(proj_langs)
+
+    def do_subproject(self, dirname):
+        subproject_dir_abs = os.path.join(self.environment.get_source_dir(), self.subproject_dir)
+        subpr = os.path.join(subproject_dir_abs, dirname)
+        try:
+            subi = BuildoptionsInterperter(subpr, '', self.backend.name, cross_file=self.cross_file, subproject=dirname, subproject_dir=self.subproject_dir, env=self.environment)
+            subi.analyze()
+        except:
+            sys.stderr.write('Parsing the subproject "{}" failed.\n'.format(dirname))
+            return
+
+    def func_add_languages(self, node, args, kwargs):
+        return self.add_languages(self.flatten_args(args))
+
+    def is_subproject(self):
+        return self.subproject != ''
+
+    def analyze(self):
+        self.load_root_meson_file()
+        self.sanity_check_ast()
+        self.parse_project()
+        self.run()
+        self.coredata.set_options(self.default_options)
```

Comment:
There are some restrictions what default options a subproject can set. I think they are restricted to setting defaults for their own user defined options. Interpreter.set_options automatically applies the subproject prefix to these options also.

Reply:
Also done. Although this had required a bit more refactoring.